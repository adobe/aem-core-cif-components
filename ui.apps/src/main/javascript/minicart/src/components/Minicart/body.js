/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';

import LoadingIndicator from '../LoadingIndicator';

import EmptyMinicartBody from './emptyMinicartBody';
import classes from './body.css';
import ProductList from './productList';
import CartOptions from './cartOptions';
import Error from './error';

import { useCartState } from './cartContext';

const loadingIndicator = <LoadingIndicator>{`Fetching cart data...`}</LoadingIndicator>;

const Body = () => {
    const [{ isEditing, cart, isLoading, errorMessage }] = useCartState();
    const isEmpty = cart && Object.entries(cart).length > 0 ? cart.items.length === 0 : true;

    if (isLoading) {
        return loadingIndicator;
    }

    if (errorMessage) {
        return <Error />;
    }

    if (isEmpty) {
        return <EmptyMinicartBody />;
    }

    if (isEditing) {
        return <CartOptions />;
    }

    const cartItems = cart.items;
    return (
        <div className={classes.root}>
            <ProductList cartItems={cartItems} />
        </div>
    );
};

export default Body;
