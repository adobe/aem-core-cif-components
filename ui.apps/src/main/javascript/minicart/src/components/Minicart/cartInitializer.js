/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import { useEffect } from 'react';
import { useCartState } from './cartContext';
import { useCookieValue } from '../../utils/hooks';
import { useMutation } from '@apollo/react-hooks';

import MUTATION_CREATE_CART from '../../queries/mutation_create_guest_cart.graphql';
import CART_DETAILS_QUERY from '../../queries/query_cart_details.graphql';
import MUTATION_REMOVE_ITEM from '../../queries/mutation_remove_item.graphql';
import MUTATION_ADD_TO_CART from '../../queries/mutation_add_to_cart.graphql';

const CartInitializer = props => {
    const [{ cartId: stateCartId }, dispatch] = useCartState();
    const CART_COOKIE = 'cif.cart';

    const [cartId, setCartCookie] = useCookieValue(CART_COOKIE);
    const [createCart, { data, error }] = useMutation(MUTATION_CREATE_CART);
    const [addItem] = useMutation(MUTATION_ADD_TO_CART);
    const [removeItem] = useMutation(MUTATION_REMOVE_ITEM);

    const createCartHandlers = (cartId, dispatch) => {
        return {
            addItem: ev => {
                if (!ev.detail) return;

                const { sku, quantity } = ev.detail;
                dispatch({ type: 'open' });
                dispatch({ type: 'beginLoading' });
                return addItem({
                    variables: { cartId, sku, quantity },
                    refetchQueries: [{ query: CART_DETAILS_QUERY, variables: { cartId } }],
                    awaitRefetchQueries: true
                })
                    .catch(error => {
                        dispatch({ type: 'error', error: error.toString() });
                    })
                    .finally(() => {
                        dispatch({ type: 'endLoading' });
                    });
            },
            removeItem: itemId => {
                dispatch({ type: 'beginLoading' });
                return removeItem({
                    variables: { cartId, itemId },
                    refetchQueries: [{ query: CART_DETAILS_QUERY, variables: { cartId } }],
                    awaitRefetchQueries: true
                })
                    .catch(error => {
                        dispatch({ type: 'error', error: error.toString() });
                    })
                    .finally(() => {
                        dispatch({ type: 'endLoading' });
                    });
            }
        };
    };

    useEffect(() => {
        if (!cartId || cartId.length === 0) {
            createCart();
        }
    }, [cartId]);

    useEffect(() => {
        if (cartId && (!stateCartId || stateCartId.length === 0)) {
            dispatch({ type: 'cartId', cartId: cartId, methods: createCartHandlers(cartId, dispatch) });
        }
    }, [cartId, stateCartId]);

    useEffect(() => {
        if (data) {
            setCartCookie(data.createEmptyCart);
            dispatch({
                type: 'cartId',
                cartId: data.createEmptyCart,
                methods: createCartHandlers(data.createEmptyCart, dispatch)
            });
        }

        if (error) {
            dispatch({ type: 'error', error: error.toString() });
        }
    }, [data, error]);

    return props.children;
};

export default CartInitializer;
