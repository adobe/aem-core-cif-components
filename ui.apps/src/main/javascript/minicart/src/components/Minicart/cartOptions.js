/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/

import React, { useState } from 'react';
import { useMutation } from '@apollo/react-hooks';
import { Price } from '@magento/peregrine';

import Button from '../Button';
import Select from '../Select';
import classes from './cartOptions.css';

import CART_DETAILS_QUERY from '../../queries/query_cart_details.graphql';
import MUTATION_UPDATE_CART_ITEM from '../../queries/mutation_update_cart_item.graphql';

import { useCartState } from './cartContext';

const CartOptions = () => {
    const [{ editItem, cartId }, dispatch] = useCartState();

    const { product, quantity: initialQuantity } = editItem;
    const { name, price: productPrice } = product;

    const { value, currency } = productPrice.regularPrice.amount;

    const [quantity, setQuantity] = useState(initialQuantity);

    const mockQtys = [
        {
            value: '1'
        },
        {
            value: '2'
        },
        {
            value: '3'
        },
        {
            value: '4'
        }
    ];

    const [updateCart] = useMutation(MUTATION_UPDATE_CART_ITEM, {
        refetchQueries: [{ query: CART_DETAILS_QUERY, variables: { cartId } }],
        awaitRefetchQueries: true
    });

    const handleUpdateClick = () => {
        updateCart({ variables: { cartId, cartItemId: editItem.id, quantity } })
            .catch(error => {
                dispatch({ type: 'error', error: error.toString() });
            })
            .finally(() => {
                dispatch({ type: 'endLoading' });
            });
        dispatch({ type: 'beginLoading' });
        dispatch({ type: 'endEditing' });
    };

    const handleOnChange = newVal => {
        setQuantity(parseInt(newVal));
    };

    return (
        <form className={classes.root}>
            <div className={classes.focusItem}>
                <span className={classes.name}>{name}</span>
                <span className={classes.price}>
                    <Price currencyCode={currency} value={value} />
                </span>
            </div>
            <div className={classes.form}>
                <section className={classes.quantity}>
                    <h2 className={classes.quantityTitle}>
                        <span>Quantity</span>
                    </h2>
                    <Select field="quantity" initialValue={quantity} onValueChange={handleOnChange} items={mockQtys} />
                </section>
            </div>
            <div className={classes.save}>
                <Button
                    onClick={() => {
                        dispatch({ type: 'endEditing' });
                    }}>
                    <span>Cancel</span>
                </Button>
                <Button priority="high" onClick={handleUpdateClick}>
                    <span>Update Cart</span>
                </Button>
            </div>
        </form>
    );
};

export default CartOptions;
