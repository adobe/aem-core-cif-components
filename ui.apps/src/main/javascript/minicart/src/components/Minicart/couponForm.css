.root {
    display: grid;
    grid-template-columns: auto 8rem;
    grid-gap: 1rem;
    margin: 0 1.5rem;
}

.root input {
    max-height: 32px;
}

.error {
    grid-column: 1 / 3;
    background-color: rgb(var(--venia-grey));
    border-radius: 4px;
    color: rgb(var(--venia-error));
    font-size: 0.875rem;
    line-height: 1.25rem;
    padding: 1rem;
}

.error:empty {
    display: none;
}
