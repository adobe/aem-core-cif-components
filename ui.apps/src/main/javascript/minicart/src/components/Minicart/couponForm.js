/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React, { useState } from 'react';
import Button from '../Button';
import { useCartState } from './cartContext';

import classes from './couponForm.css';

const CouponForm = () => {
    const [{ addCoupon, couponError }] = useCartState();
    const [couponCode, setCouponCode] = useState('');

    const addCouponHandler = () => {
        return addCoupon(couponCode);
    };

    const errorFragment = couponError ? <div className={classes.error}>{couponError}</div> : '';

    return (
        <form className={classes.root}>
            <input
                value={couponCode}
                onChange={e => setCouponCode(e.target.value)}
                type="text"
                name="couponCode"
                placeholder="Enter your code"
            />
            <Button priority="high" onClick={addCouponHandler} disabled={couponCode.length < 3}>
                <span>Apply Coupon</span>
            </Button>
            {errorFragment}
        </form>
    );
};

export default CouponForm;
