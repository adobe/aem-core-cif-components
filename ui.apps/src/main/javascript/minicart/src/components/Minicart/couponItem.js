/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { useCartState } from './cartContext';
import { useMutation } from '@apollo/react-hooks';
import Kebab from './kebab';
import Section from './section';

import classes from './couponItem.css';

import MUTATION_REMOVE_COUPON from '../../queries/mutation_remove_coupon.graphql';
import CART_DETAILS_QUERY from '../../queries/query_cart_details.graphql';

const CouponItem = () => {
    const [{ cartId, cart }, dispatch] = useCartState();

    const [removeCoupon] = useMutation(MUTATION_REMOVE_COUPON);

    const removeCouponHandler = () => {
        dispatch({ type: 'beginLoading' });
        return removeCoupon({
            variables: { cartId },
            refetchQueries: [{ query: CART_DETAILS_QUERY, variables: { cartId } }],
            awaitRefetchQueries: true
        }).finally(() => {
            dispatch({ type: 'endLoading' });
        });
    };

    const appliedCoupon = cart.applied_coupon ? cart.applied_coupon.code : null;

    return (
        <div className={classes.root}>
            <div className={classes.couponName}>
                Coupon <strong>{appliedCoupon}</strong> applied.
            </div>
            <Kebab>
                <Section text="Remove coupon" onClick={removeCouponHandler} icon="Trash" />
            </Kebab>
        </div>
    );
};

export default CouponItem;
