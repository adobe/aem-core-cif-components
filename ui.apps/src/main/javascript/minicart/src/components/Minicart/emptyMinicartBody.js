/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { func } from 'prop-types';
import Trigger from '../Trigger';
import classes from './emptyMiniCartBody.css';

const EmptyMinicartBody = props => {
    const { closeDrawer } = props;
    return (
        <div className={classes.root} data-testid="empty-minicart">
            <h3 className={classes.emptyTitle}>There are no items in your shopping cart</h3>
            <Trigger action={closeDrawer}>
                <span className={classes.continue}>Continue Shopping</span>
            </Trigger>
        </div>
    );
};

EmptyMinicartBody.propTypes = {
    closeDrawer: func.isRequired
};

export default EmptyMinicartBody;
