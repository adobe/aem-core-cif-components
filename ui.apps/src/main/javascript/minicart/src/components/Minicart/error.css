.root {
    background-color: rgb(var(--venia-warning-light));
    box-shadow: 0 1px rgb(var(--venia-border));
    margin: 10px;
    padding: 10px;
}

.root h2 {
    color: rgb(var(--venia-warning-dark));
    padding-bottom: 0.83rem;
}
