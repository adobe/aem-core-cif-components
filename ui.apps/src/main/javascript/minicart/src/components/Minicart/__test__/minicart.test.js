/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import ReactDOM from 'react-dom';
import { render } from 'test-utils';

import Minicart from '../minicart';
import { waitForElement } from '@testing-library/dom';
import { CartProvider } from '../cartContext';

describe('<Minicart>', () => {
    beforeAll(() => {
        // mock createPortal because we don't have the DOM element to render the CartTrigger
        ReactDOM.createPortal = jest.fn(element => {
            return element;
        });
    });

    it('renders the empty cart', async () => {
        const { getByTestId } = render(
            <CartProvider initialState={{ cartId: 'empty' }} reducerFactory={() => state => state}>
                <Minicart />
            </CartProvider>
        );

        // the component is rendered async (the "Fetching cart data is displayed on first render") so we await the element to be ready
        // getByTestId() throws an error if the element will not be available.
        const emptyCartNode = await waitForElement(() => getByTestId('empty-minicart'));

        // compare the snapshow of the element with the stored one.
        expect(emptyCartNode).toMatchSnapshot();
    });
});
