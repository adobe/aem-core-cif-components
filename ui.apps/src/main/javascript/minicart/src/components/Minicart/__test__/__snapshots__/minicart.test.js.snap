// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Minicart> renders the empty cart 1`] = `
<div
  class="root"
  data-testid="empty-minicart"
>
  <h3
    class="emptyTitle"
  >
    There are no items in your shopping cart
  </h3>
  <button
    class="root"
    type="button"
  >
    <span
      class="continue"
    >
      Continue Shopping
    </span>
  </button>
</div>
`;
